public/framework/catalog_cache/cache_manager.ts

```diff
@@ -0,0 +1,204 @@
+  }
+
+  /**
+   * Retrieves accelerations cache from local storage.
+   * @returns {AccelerationsCacheData} The retrieved accelerations cache.
+   */
+  static getAccelerationsCache(): AccelerationsCacheData {
+    const accelerationCacheData = localStorage.getItem(this.accelerationsCacheKey);
+
+    if (accelerationCacheData) {
+      return JSON.parse(accelerationCacheData);
+    } else {
+      const defaultCacheObject = {
+        version: '1.0',
+        accelerations: [],
+        lastUpdated: '',
+        status: CachedDataSourceStatus.Empty,
+      };
+      this.saveAccelerationsCache(defaultCacheObject);
+      return defaultCacheObject;
+    }
+  }
+
+  /**
+   * Adds or updates a data source in the cache.
+   * @param {CachedDataSource} dataSource - The data source to add or update.
+   */
+  static addOrUpdateDataSource(dataSource: CachedDataSource): void {
+    const cacheData = this.getDataSourceCache();
+    const index = cacheData.dataSources.findIndex(
+      (ds: CachedDataSource) => ds.name === dataSource.name
+    );
+    if (index !== -1) {
+      cacheData.dataSources[index] = dataSource;
+    } else {
+      cacheData.dataSources.push(dataSource);
+    }
+    this.saveDataSourceCache(cacheData);
+  }
+
+  /**
+   * Retrieves or creates a data source with the specified name.
+   * @param {string} dataSourceName - The name of the data source.
+   * @returns {CachedDataSource} The retrieved or created data source.
+   */
+  static getOrCreateDataSource(dataSourceName: string): CachedDataSource {
+    const cacheData = this.getDataSourceCache();
+    const cachedDataSource = cacheData.dataSources.find(
+      (ds: CachedDataSource) => ds.name === dataSourceName
+    );
+    if (cachedDataSource) {
+      return cachedDataSource;
+    } else {
+      const defaultDataSourceObject = {
+        name: dataSourceName,
+        lastUpdated: '',
+        status: CachedDataSourceStatus.Empty,
+        databases: [],
+      };
+      this.addOrUpdateDataSource(defaultDataSourceObject);
+      return defaultDataSourceObject;
+    }
+  }
+
+  /**
+   * Retrieves a database from the cache.
+   * @param {string} dataSourceName - The name of the data source containing the database.
+   * @param {string} databaseName - The name of the database.
+   * @returns {CachedDatabase} The retrieved database.
+   * @throws {Error} If the data source or database is not found.
+   */
+  static getDatabase(dataSourceName: string, databaseName: string): CachedDatabase {
+    const cachedDataSource = this.getDataSourceCache().dataSources.find(
+      (ds) => ds.name === dataSourceName
+    );
+    if (!cachedDataSource) {
+      throw new Error('DataSource not found exception: ' + dataSourceName);
+    }
+
+    const cachedDatabase = cachedDataSource.databases.find((db) => db.name === databaseName);
+    if (!cachedDatabase) {
+      throw new Error('Database not found exception: ' + databaseName);
```

Comment:
From an API usability standpoint, I wonder if it's good design to throw an error on a cache miss. Since it's a pretty regular and probably expected occurrence, it might make more sense to return a `CachedDatabase | null` (or reuse the `ok: boolean` pattern from `AsyncApiResponse`) so TypeScript can do more heavy lifting for soundness checking.

Reply:
This is by design right now. We want the cache users to know where the get request is failing. I'll check with @RyanL1997 @sejli what do they expect `thrown error or an error response`. I remember UI having special cases where they see something missing with toast errors or callouts.